File: csharp/AppEncryption/AppEncryption.Tests/AppEncryption/Persistence/PersistenceTest.cs
using System;
using GoDaddy.Asherah.AppEncryption.Persistence;
using Moq;
using Xunit;

namespace GoDaddy.Asherah.AppEncryption.Tests.AppEncryption.Persistence
{
    [Collection("Logger Fixture collection")]
    public class PersistenceTest
    {
        private readonly Mock<Persistence<string>> persistenceMock;

        public PersistenceTest()
        {
            persistenceMock = new Mock<Persistence<string>>();
        }

        [Fact]
        private void TestStore()
        {
            const string expectedPersistenceKey = "some_key";
            persistenceMock.Setup(x => x.GenerateKey(It.IsAny<string>())).Returns(expectedPersistenceKey);
            persistenceMock.Setup(x => x.Store(It.IsAny<string>(), It.IsAny<string>())).Verifiable();
            persistenceMock.Setup(x => x.Store(It.IsAny<string>())).CallBase();

            const string value = "some_value";
            string actualPersistenceKey = persistenceMock.Object.Store(value);
            Assert.Equal(expectedPersistenceKey, actualPersistenceKey);
            persistenceMock.Verify(x => x.GenerateKey(value));
            persistenceMock.Verify(x => x.Store(expectedPersistenceKey, value));
        }

        [Fact]
        private void TestGenerateKey()
        {
            const string value = "unused_value";
            persistenceMock.Setup(x => x.GenerateKey(It.IsAny<string>())).CallBase();

            string returnedValue = persistenceMock.Object.GenerateKey(value);

            // Just verify it's a valid UUID
            Assert.True(Guid.TryParse(returnedValue, out Guid guidOutput));
        }
    }
}

File: java/app-encryption/src/test/java/com/godaddy/asherah/appencryption/persistence/PersistenceTest.java
package com.godaddy.asherah.appencryption.persistence;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.Mockito.*;

import java.util.UUID;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

@ExtendWith(MockitoExtension.class)
class PersistenceTest {
  @Mock
  Persistence<String> persistence;

  @Test
  void testStore() {
    String expectedPersistenceKey = "some_key";
    when(persistence.generateKey(any())).thenReturn(expectedPersistenceKey);
    doNothing().when(persistence).store(anyString(), anyString());
    doCallRealMethod().when(persistence).store(anyString());

    String value = "some_value";
    String actualPersistenceKey = persistence.store(value);
    assertEquals(expectedPersistenceKey, actualPersistenceKey);
    verify(persistence).generateKey(value);
    verify(persistence).store(eq(expectedPersistenceKey), eq(value));
  }

  @Test
  void testGenerateKey() {
    String value= "unused_value";
    when(persistence.generateKey(anyString())).thenCallRealMethod();

    String returnedValue = persistence.generateKey(value);
    // Just verify it's a valid UUID (note JDK8 has known bug https://bugs.java.com/view_bug.do?bug_id=8159339)
    UUID.fromString(returnedValue);
  }

}
